src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -0,0 +1,346 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+import is.hail.annotations.MemoryBuffer
+import is.hail.asm4s.FunctionBuilder.methodNodeToGrowable
+import is.hail.asm4s._
+import is.hail.expr
+import is.hail.expr.{TInt32, TInt64, TArray, TContainer, TStruct, TFloat32, TFloat64, TBoolean}
+import is.hail.annotations.StagedRegionValueBuilder
+import scala.collection.generic.Growable
+
+import org.objectweb.asm.Opcodes._
+import org.objectweb.asm.Type
+import org.objectweb.asm.tree._
+
+import scala.reflect.classTag
+import scala.reflect.ClassTag
+
+object Compile {
+  private def dummyValue(t: expr.Type): Code[_] = t match {
+    case TBoolean => false
+    case TInt32 => 0
+    case TInt64 => 0L
+    case TFloat32 => 0.0f
+    case TFloat64 => 0.0
+    case _ => 0L // reference types
+  }
+  private def typeToTypeInfo(t: expr.Type): TypeInfo[_] = t match {
+    case TInt32 => typeInfo[Int]
+    case TInt64 => typeInfo[Long]
+    case TFloat32 => typeInfo[Float]
+    case TFloat64 => typeInfo[Double]
+    case TBoolean => typeInfo[Boolean]
+    case _ => typeInfo[Long] // reference types
+  }
+
+  type E = Env[(TypeInfo[_], Code[Boolean], Code[_])]
+
+  def apply(ir: IR, fb: FunctionBuilder[_]) {
+    apply(ir, fb, new Env())
+  }
+
+  def apply(ir: IR, fb: FunctionBuilder[_], env: E) {
+    fb.emit(expression(ir, fb, env, new StagedBitSet(fb))._2)
```

Comment:
I feel like the return value should be ignored at the top level.

Reply:
I don't understand. To which return value are you referring? I ignore the missingness here. If I ignore the return value of `expression` then I won't emit the `Code[_]` corresponding to the bulk of the computation.

If you're referencing the fact that `Out` is unnecessary, I think I agree. I added `Out` as a way to possibly return multiple values, but I think IR expressions always evaluate to a single value and have no side-effects, so I'm inclined to remote `Out` (equivalently: there's always an implicit `Out` around the top of every expression).